src/slice-assign.c

```diff
@@ -52,39 +63,71 @@ SEXP vec_assign(SEXP x, SEXP index, SEXP value,
+  proxy = PROTECT(vec_proxy_assign_opts(proxy, index, value, opts));
 
   SEXP out = vec_restore(proxy, x, R_NilValue);
 
   UNPROTECT(6);
   return out;
 }
 
-SEXP vec_assign_switch(SEXP proxy, SEXP index, SEXP value) {
+// [[ register() ]]
+SEXP vctrs_assign_params(SEXP x, SEXP index, SEXP value,
+                         SEXP assign_names) {
+  struct vec_assign_opts opts = {
+    .assign_names = r_bool_as_int(assign_names),
+    .x_arg = args_empty,
+    .value_arg = args_empty
+  };
+  return vec_assign_opts(x, index, value, &opts);
+}
+
+static SEXP vec_assign_switch(SEXP proxy, SEXP index, SEXP value,
+                              const struct vec_assign_opts* opts) {
   switch (vec_proxy_typeof(proxy)) {
-  case vctrs_type_logical:     return lgl_assign(proxy, index, value);
-  case vctrs_type_integer:     return int_assign(proxy, index, value);
-  case vctrs_type_double:      return dbl_assign(proxy, index, value);
-  case vctrs_type_complex:     return cpl_assign(proxy, index, value);
-  case vctrs_type_character:   return chr_assign(proxy, index, value);
-  case vctrs_type_raw:         return raw_assign(proxy, index, value);
-  case vctrs_type_list:        return list_assign(proxy, index, value);
-  case vctrs_type_dataframe:   return df_assign(proxy, index, value);
-  case vctrs_type_null:
-  case vctrs_type_unspecified:
-  case vctrs_type_s3:
-                               Rf_error("Internal error in `vec_assign_switch()`: Unexpected type %s.",
-                                        vec_type_as_str(vec_typeof(proxy)));
-  case vctrs_type_scalar:      stop_scalar_type(proxy, args_empty);
+  case vctrs_type_logical:   return lgl_assign(proxy, index, value);
+  case vctrs_type_integer:   return int_assign(proxy, index, value);
+  case vctrs_type_double:    return dbl_assign(proxy, index, value);
+  case vctrs_type_complex:   return cpl_assign(proxy, index, value);
+  case vctrs_type_character: return chr_assign(proxy, index, value);
+  case vctrs_type_raw:       return raw_assign(proxy, index, value);
+  case vctrs_type_list:      return list_assign(proxy, index, value);
+  case vctrs_type_dataframe: return df_assign(proxy, index, value, opts);
+  case vctrs_type_scalar:    stop_scalar_type(proxy, args_empty);
+  default:                   vctrs_stop_unsupported_type(vec_typeof(proxy), "vec_assign_switch()");
   }
   never_reached("vec_assign_switch");
 }
 
+SEXP vec_proxy_assign_names(SEXP proxy, SEXP index, SEXP value) {
+  SEXP value_nms = PROTECT(vec_names(value));
+
+  if (value_nms == R_NilValue) {
+    UNPROTECT(1);
+    return proxy;
+  }
+
+  SEXP proxy_nms = PROTECT(vec_names(proxy));
+  if (proxy_nms == R_NilValue) {
+    proxy_nms = Rf_allocVector(STRSXP, vec_size(proxy));
+    UNPROTECT(1);
+    PROTECT(proxy_nms);
+  }
+
+  proxy_nms = PROTECT(chr_assign(proxy_nms, index, value_nms));
+
+  proxy = PROTECT(r_maybe_duplicate(proxy));
+  vec_set_names(proxy, proxy_nms);
```

Comment:
`vec_set_names()` works like `vec_proxy_assign()`. It might have to call back to `names<-` and duplicate the input, so it always returns it's input. So you won't need to duplicate the proxy, but I think you do need to do `proxy = PROTECT(vec_set_names())`